robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java

```diff
@@ -682,16 +688,56 @@ private String createTempDir(String name) {
         .toString();
   }
 
+  private static BroadcastReceiver newBroadcastReceiverFromP(
+      String receiverClassName,
+      Application application,
+      LoadedApk loadedApk,
+      BroadcastReceiverData receiver) {
+    ClassLoader classLoader;
+    if (application.getBaseContext() != null) {
+      classLoader = application.getBaseContext().getClassLoader();
+    } else {
+      classLoader = loadedApk.getClassLoader();
+    }
+    for (String action : receiver.getActions()) {
+      try {
+        // If one BroadcastReceiver has multiple actions, we will try to load it with action one
+        // by one until we can initialize this BroadcastReceiver correctly.
+        Intent intent = new Intent(action);
+        BroadcastReceiver broadcastReceiver =
+            loadedApk.getAppFactory().instantiateReceiver(classLoader, receiverClassName, intent);
+        if (broadcastReceiver == null) {
+          System.err.println(
+              "Failed to initialize receiver "
+                  + receiverClassName
+                  + " with AppComponentFactory "
+                  + loadedApk.getAppFactory());
+        } else {
+          return broadcastReceiver;
+        }
+      } catch (ClassNotFoundException | IllegalAccessException | InstantiationException ignored) {
+        // Ignore these errors
+      }
+    }
+    return null;
+  }
+
   // TODO move/replace this with packageManager
   @VisibleForTesting
-  static void registerBroadcastReceivers(Application application, AndroidManifest androidManifest) {
+  static void registerBroadcastReceivers(
+      Application application, AndroidManifest androidManifest, LoadedApk loadedApk) {
     for (BroadcastReceiverData receiver : androidManifest.getBroadcastReceivers()) {
       IntentFilter filter = new IntentFilter();
       for (String action : receiver.getActions()) {
         filter.addAction(action);
       }
       String receiverClassName = replaceLastDotWith$IfInnerStaticClass(receiver.getName());
-      application.registerReceiver((BroadcastReceiver) newInstanceOf(receiverClassName), filter);
+      if (loadedApk != null && Build.VERSION.SDK_INT >= P) {
```

Comment:
`RuntimeEnvironment.getApiLevel()` is preferable to `Build.VERSION.SDK_INT`, because sadly it's common for people to modify the return value of `Build.VERSION.SDK_INT`.

Reply:
Thanks for pointing it out. It's my first time to know this usage of `SDK_INT`.